mpqp/execution/providers/google.py

```
@@ -18,6 +20,96 @@
+    nb_qubits: int,
+) -> "Circuit":
+    """Apply noise models to a Cirq circuit.
+
+    This function applies noise models to a given Cirq circuit based on the specified noise models and
+    the number of qubits in the circuit. It modifies the original circuit by adding noise operations
+    after the original gates and returns a new circuit with the noise applied.
+
+    Args:
+        cirq_circuit: The Cirq circuit to apply noise to.
+        noises: A list of noise models to apply to the circuit.
+        nb_qubits: The number of qubits in the circuit.
+
+    Returns:
+        A new circuit with the noise operations applied.
+    """
+    from cirq.ops.identity import IdentityGate
+    from cirq.ops.measurement_gate import MeasurementGate
+    from cirq.ops.raw_types import Gate, Operation
+
+    qubit_list = sorted(cirq_circuit.all_qubits())
+    noisy_moments = []
+
+    for moment in cirq_circuit:
+        moment_ops = list(moment.operations)
+        noisy_moments.append(Moment(moment_ops))
+
+        qubit_noise_op: list[list[Operation]] = [[] for _ in range(nb_qubits)]
+
+        for op in moment_ops:
+            if isinstance(op.gate, (MeasurementGate, IdentityGate)):
+                continue
+
+            for noise in reversed(noises):
+                cirq_noise = noise.to_other_language(Language.CIRQ)
+                if TYPE_CHECKING:
+                    assert isinstance(cirq_noise, Gate)
+
+                noise_dimension = getattr(noise, "dimension", 1)
+
+                if not noise.targets or len(noise.targets) == nb_qubits:
+                    target_qubits = qubit_list
+                else:
+                    target_qubits = [qubit_list[i] for i in noise.targets]
+
+                allowed_gates = set()
+                if noise.gates:
+                    for gate in noise.gates:
+                        gate_obj = (
+                            gate.cirq_gate if hasattr(gate, "cirq_gate") else gate
+                        )
```

I am not fan of the `else gate`. 
Each NativeGate should have an attribute `cirq_gate`, so if the user has put a funky NativeGate in `noise.gates` and forgot to implement the property `cirq_gate` we shouldn't cover his back here, we should raise an error.

Also here putting `gate` will make gate_obj being either a cirq or mpqp gate, which is a bit weird

---

Here 50a7425